### PR TITLE
User Story #16.3: Activate remedial actions only after a specific constraint

### PR DESCRIPTION
**Pull request by pet-mit**
**Please check if the PR fulfills these requirements** *(please use `'[x]'` to check the checkboxes, or submit the PR and then click the checkboxes)*
- [ ] The commit message follows our guidelines
- [ ] Tests for the changes have been added (for bug fixes / features)
- [ ] Docs have been added / updated (for bug fixes / features)

**What kind of change does this PR introduce?** *(Bug fix, feature, docs update, ...)*
Feature

**What is the current behavior?** *(You can also link to an open issue here)*
availability of remedial actions is assessed at the beginning of the search tree, regardless of the state of the network

**What is the new behavior (if this is a feature change)?**
availability of remedial actions is reassessed during the search tree, in order to be able to activate some remedial actions only after a specific constraint on an element

**Does this PR introduce a breaking change?** *(What changes might users need to make in their application due to this PR?)*
Some changes in the Crac API, with the creation of the OnFlowConstraint UsageRule implementation

### CODE REVIEW

**Review by phiedw — CHANGES_REQUESTED** [review]

File: data/crac/crac-api/src/main/java/com/farao_community/farao/data/crac_api/Crac.java [hunk 1]
```diff
@@ -270,7 +270,7 @@ default State getState(String contingencyId, Instant instant) {
      * Gather all the range actions of a specified state with the specified usage method (available, forced or
```
[the remark this answers is not in the document]
Comment (pet-mit) [reply data/crac/crac-api/src/main/java/com/farao_community/farao/data/crac_api/Crac.java:270]: done

File: data/crac/crac-impl/src/main/java/com/farao_community/farao/data/crac_impl/OnFlowConstraintImpl.java [hunk 2]
```diff
@@ -0,0 +1,59 @@
...
+    }
+
+    @Override
+    public Instant getInstant() {
+        return instant;
+    }
+
+    @Override
+    public UsageMethod getUsageMethod(State state) {
+        return state.getInstant().equals(instant) ? UsageMethod.TO_BE_EVALUATED : UsageMethod.UNDEFINED;
```
[the remark this answers is not in the document]
Comment (pet-mit) [reply data/crac/crac-impl/src/main/java/com/farao_community/farao/data/crac_impl/OnFlowConstraintImpl.java:40]: unavailable would mean that the RA is definitely unavailable, no matter the other usage rules (UNAVAILABLE > other usage methods).
undefined means that this specific usage rule cannot say if the RA is available or not.

File: ra-optimisation/search-tree-rao/src/main/java/com/farao_community/farao/search_tree_rao/RangeActionFilter.java [hunk 3]
```diff
@@ -132,13 +146,29 @@ public void filterMaxRas() {
...
+                .sorted((ra1, ra2) -> comparePrioritiesAndSensitivities(ra1, ra2, leaf.getMostLimitingElements(1).get(0), leaf))
+                .collect(Collectors.toList());
             rangeActionsToRemove.removeAll(rangeActionsSortedBySensitivity.subList(0, updatedNumberOfRangeActionsToKeep));
             return rangeActionsToRemove;
         }
     }
 
+    /**
+     * First compares priority then sensi
+     * If a range action has more priority (depending on the contents of leastPriorityRangeActions) then the other, then
```
[the remark this answers is not in the document]
Comment (pet-mit) [reply ra-optimisation/search-tree-rao/src/main/java/com/farao_community/farao/search_tree_rao/RangeActionFilter.java]: corrected

File: data/crac/crac-impl/src/main/java/com/farao_community/farao/data/crac_impl/OnFlowConstraintImpl.java (hunk 2 above)
[the remark this answers is not in the document]
Comment (phiedw) [reply data/crac/crac-impl/src/main/java/com/farao_community/farao/data/crac_impl/OnFlowConstraintImpl.java:40]: Indeed that makes sense :+1:

**Review by phiedw — APPROVED** [review]

**Pull request merged**